### PR TITLE
Handle canary failure case

### PR DESCRIPTION
**Pull request by zanthrash**
We were not handling the FAILURE case for canaries. This was causing an IllegialStateException to be thrown instead fo setting the stage to TERMINAL or FAILED_CONTINUE if 'continueOnUnhealthy'  is true

### CODE REVIEW

File: orca-mine/src/main/groovy/com/netflix/spinnaker/orca/mine/tasks/CompleteCanaryTask.groovy [hunk 1]
```diff
@@ -32,10 +32,12 @@ class CompleteCanaryTask implements Task {
...
     } else if (canary.canaryResult?.overallResult == 'SUCCESS') {
       return new DefaultTaskResult(ExecutionStatus.SUCCEEDED)
-    } else if (stage.context.continueOnUnhealthy && canary.health.health in ["UNHEALTHY", "UNKNOWN"]) {
+    } else if (stage.context.continueOnUnhealthy == true && (canary?.health?.health in ["UNHEALTHY", "UNKNOWN"] || canary.canaryResult?.overallResult == 'FAILURE') ) {
       return new DefaultTaskResult(ExecutionStatus.FAILED_CONTINUE)
+    } else if (canary.canaryResult?.overallResult == 'FAILURE') {
+      return new DefaultTaskResult(ExecutionStatus.TERMINAL)
     } else {
-      throw new IllegalStateException("Canary failed")
+      throw new IllegalStateException("Canary failed: [overallResult: ${canary?.canaryResult?.overallResult}, health: ${canary?.health?.health}, status: ${canary?.status?.status}, continueOnUnhealthy: ${canary?.context?.continueOnUnhealthy}]")
```
Comment (ajordens) [inline orca-mine/src/main/groovy/com/netflix/spinnaker/orca/mine/tasks/CompleteCanaryTask.groovy]: Does this make it's way to the UI as an error message or anything?

If we're looking at this in logs, I'd suggest including the `executionId` (perhaps more valuable than the other bits as they should be included in the context).

File: orca-mine/src/main/groovy/com/netflix/spinnaker/orca/mine/tasks/CompleteCanaryTask.groovy [hunk 2]
```diff
@@ -32,10 +32,12 @@ class CompleteCanaryTask implements Task {
       return new DefaultTaskResult(ExecutionStatus.CANCELED)
     } else if (canary.canaryResult?.overallResult == 'SUCCESS') {
       return new DefaultTaskResult(ExecutionStatus.SUCCEEDED)
-    } else if (stage.context.continueOnUnhealthy && canary.health.health in ["UNHEALTHY", "UNKNOWN"]) {
+    } else if (stage.context.continueOnUnhealthy == true && (canary?.health?.health in ["UNHEALTHY", "UNKNOWN"] || canary.canaryResult?.overallResult == 'FAILURE') ) {
```
Comment (ajordens) [inline orca-mine/src/main/groovy/com/netflix/spinnaker/orca/mine/tasks/CompleteCanaryTask.groovy]:
```

File: orca-mine/src/main/groovy/com/netflix/spinnaker/orca/mine/tasks/CompleteCanaryTask.groovy (hunk 1 above)
Comment (zanthrash) [inline orca-mine/src/main/groovy/com/netflix/spinnaker/orca/mine/tasks/CompleteCanaryTask.groovy]: Good point

File: orca-mine/src/main/groovy/com/netflix/spinnaker/orca/mine/tasks/CompleteCanaryTask.groovy (hunk 2 above)
Comment (zanthrash) [inline orca-mine/src/main/groovy/com/netflix/spinnaker/orca/mine/tasks/CompleteCanaryTask.groovy]: Agreed

**Pull request merged**